core/trino-main/src/test/java/io/trino/operator/TestGroupByHash.java

```diff
@@ -502,4 +505,119 @@ public void testMemoryReservationYieldWithDictionary()
         assertEquals(currentQuota.get(), 10);
         assertEquals(currentQuota.get() / 3, yields);
     }
+
+    @Test
+    public void testLowCardinalityDictionariesAddPage()
+    {
+        GroupByHash groupByHash = createGroupByHash(
+                TEST_SESSION,
+                ImmutableList.of(BIGINT, BIGINT),
+                new int[] {0, 1},
+                Optional.empty(),
+                100,
+                JOIN_COMPILER,
+                TYPE_OPERATOR_FACTORY,
+                NOOP);
+        Block firstBlock = BlockAssertions.createLongDictionaryBlock(0, 100, 10);
+        Block secondBlock = BlockAssertions.createLongDictionaryBlock(0, 100, 10);
+        Page page = new Page(firstBlock, secondBlock);
+
+        groupByHash.addPage(page).process();
+        assertThat(groupByHash.getGroupCount()).isEqualTo(10); // Blocks are identical so only 10 distinct groups
+
+        firstBlock = BlockAssertions.createLongDictionaryBlock(10, 100, 5);
+        secondBlock = BlockAssertions.createLongDictionaryBlock(10, 100, 7);
+        page = new Page(firstBlock, secondBlock);
+
+        groupByHash.addPage(page).process();
+        assertThat(groupByHash.getGroupCount()).isEqualTo(45); // Old 10 groups and 35 new
+    }
+
+    @Test
+    public void testCompareLowCardinalityWithGetGroupIds()
+    {
+        GroupByHash groupByHash = createGroupByHash(
+                TEST_SESSION,
+                ImmutableList.of(BIGINT, BIGINT, BIGINT, BIGINT, BIGINT),
+                new int[] {0, 1, 2, 3, 4},
+                Optional.empty(),
+                100,
+                JOIN_COMPILER,
+                TYPE_OPERATOR_FACTORY,
+                NOOP);
+
+        GroupByHash lowCardinalityGroupByHash = createGroupByHash(
+                TEST_SESSION,
+                ImmutableList.of(BIGINT, BIGINT, BIGINT, BIGINT),
+                new int[] {0, 1, 2, 3},
+                Optional.empty(),
+                100,
+                JOIN_COMPILER,
+                TYPE_OPERATOR_FACTORY,
+                NOOP);
+        Block sameValueBlock = BlockAssertions.createLongRepeatBlock(0, 100);
+        Block block1 = BlockAssertions.createLongDictionaryBlock(0, 100, 1);
+        Block block2 = BlockAssertions.createLongDictionaryBlock(0, 100, 2);
+        Block block3 = BlockAssertions.createLongDictionaryBlock(0, 100, 3);
+        Block block4 = BlockAssertions.createLongDictionaryBlock(0, 100, 4);
+        // Combining block 2 and 4 will result in only 4 distinct values since 2 and 4 are not coprime
+
+        Page lowCardinalityPage = new Page(block1, block2, block3, block4);
+        Page page = new Page(block1, block2, block3, block4, sameValueBlock); // sameValueBlock will prevent low cardinality optimization to fire
+        Work<GroupByIdBlock> lowCardinalityWork = lowCardinalityGroupByHash.getGroupIds(lowCardinalityPage);
```

Comment:
It's better to test type here instead of `testProperWorkTypesSelected` because then we are certain that page is processed as we expected